pallets/foreign-investments/src/impls/redeem.rs

```diff
@@ -584,3 +584,137 @@ where
 		}
 	}
 }
+
+#[cfg(test)]
+mod tests {
+	use rand::{rngs::StdRng, seq::SliceRandom, SeedableRng};
+
+	use super::*;
+
+	#[derive(Clone, Copy, PartialEq, Debug)]
+	enum CurrencyId {
+		Foreign,
+		Pool,
+	}
+
+	type RedeemState = super::RedeemState<u128, CurrencyId>;
+	type RedeemTransition = super::RedeemTransition<u128, CurrencyId>;
+
+	impl RedeemState {
+		fn get_done_amount(&self) -> u128 {
+			match *self {
+				Self::ActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+					done_amount, ..
+				} => done_amount,
+				Self::RedeemingAndActiveSwapIntoForeignCurrencyAndSwapIntoForeignDone {
+					done_amount,
+					..
+				} => done_amount,
+				Self::SwapIntoForeignDone { done_swap } => done_swap.amount,
+				Self::RedeemingAndSwapIntoForeignDone { done_swap, .. } => done_swap.amount,
+				_ => 0,
+			}
+		}
+
+		fn get_swap_amount(&self) -> u128 {
+			self.get_active_swap().map(|swap| swap.amount).unwrap_or(0)
+		}
+
+		fn total(&self) -> u128 {
+			self.get_redeeming_amount() + self.get_done_amount() + self.get_swap_amount()
+		}
+	}
+
+	struct Checker {
```

Comment:
This looks nice, maybe in the future we can make this a bit more generic so we can use it in other places as well.

Reply:
Yeah! As a starting point can probably be factorized between `Invest` & `Checker` types. 
BTW, referring to generic fuzzer. SRLabs gives a talk in Sub0 about its fuzzer which is now open: https://github.com/srlabs/substrate-runtime-fuzzer

It could be interesting to use it. I liked the talk if you have the opportunity to see it.

Reply:
I need to rewatch it with full attention once the VODs are out but it was super interesting!